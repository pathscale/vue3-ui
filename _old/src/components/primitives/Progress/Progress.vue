<script>
import { computed, watchEffect, nextTick, ref } from 'vue'

export default {
  name: 'VProgress',
  props: {
    type: {
      type: [String, Object],
      default: 'is-darkgrey'
    },
    size: String,
    value: Number,
    max: {
      type: Number,
      default: 100
    },
    showValue: Boolean,
    format: {
      type: String,
      default: 'raw'
    },
    precision: {
      type: Number,
      default: 2
    },
    keepTrailingZeroes: Boolean
  },
  setup(props, { slot }) {
    const progress = ref(null)

    const toFixed = num => {
      let fixed = (Number(`${Math.round(Number(`${num}e${props.precision}`))}e${-props.precision}`)).toFixed(props.precision)
      if (!props.keepTrailingZeroes) {
        fixed = fixed.replace(/\.?0+$/, '')
      }
      return fixed
    }

    watchEffect(() =>
      nextTick(() => {
        if (isIndeterminate.value) {
          progress.value.removeAttribute('value')
        } else {
          progress.value.setAttribute('value', props.value)
        }
      })
    )

    const isIndeterminate = computed(() => {
      return props.value === undefined || props.value === null
    })

    const beyondHalf = computed(() => props.value >= 50 )

    const newValue = computed(() => {
      if (props.value === undefined || props.value === null || Number.isNaN(props.value)) {
        return undefined
      }

      if (props.format === 'percent') {
        const val = toFixed(props.value * 100 / props.max)
        return `${val}%`
      }
      const val = toFixed(props.value)
      return val
    })

    return { beyondHalf, newValue, progress }
  }
}
</script>

<template>
  <div class="progress-wrapper">
    <progress
      ref="progress"
      class="progress"
      :class="[size, type]"
      :max="max"
      :value="value">
      {{ newValue }}
    </progress>
    <p
      v-if="showValue"
      class="progress-value"
      :class="{ 'more-than-half': beyondHalf }">
      <slot>{{ newValue }}</slot>
    </p>
  </div>
</template>
