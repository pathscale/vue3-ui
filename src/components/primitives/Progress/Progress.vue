<script setup lang="ts">
import { formatToFixed } from "@/utils/functions";
import { computed, nextTick, useTemplateRef, watchEffect } from "vue";

type ClassValue = string | string[] | Record<string, boolean>;

const props = withDefaults(
  defineProps<{
    type?:
      | "is-white"
      | "is-light"
      | "is-dark"
      | "is-black"
      | "is-text"
      | "is-primary"
      | "is-link"
      | "is-info"
      | "is-success"
      | "is-warning"
      | "is-danger"
      | "is-darkgrey";
    size?: "is-small" | "is-medium" | "is-large";
    value?: number | null | undefined;
    max?: number;
    showValue?: boolean;
    format?: "raw" | "percent";
    precision?: number;
    keepTrailingZeroes?: boolean;
  }>(),
  {
    type: "is-darkgrey",
    max: 100,
    format: "raw",
    precision: 2,
  },
);

const progress = useTemplateRef("progress");

watchEffect(() =>
  nextTick(() => {
    if (isIndeterminate.value) {
      progress.value.removeAttribute("value");
    } else {
      progress.value.setAttribute("value", props.value);
    }
  }),
);

const isIndeterminate = computed(() => {
  return props.value === undefined || props.value === null;
});

const beyondHalf = computed(() => props.value >= 50);

const newValue = computed(() => {
  if (
    props.value === undefined ||
    props.value === null ||
    Number.isNaN(props.value)
  ) {
    return undefined;
  }

  // 'percent' format
  if (props.format === "percent") {
    const val = formatToFixed(
      (props.value * 100) / props.max,
      props.precision,
      props.keepTrailingZeroes,
    );
    return `${val}%`;
  }

  // 'raw' format
  return formatToFixed(props.value, props.precision, props.keepTrailingZeroes);
});
</script>

<template>
  <div class="progress-wrapper">
    <progress
      ref="progress"
      class="progress"
      :class="[size, type]"
      :max="max"
      :value="value">
      {{ newValue }}
    </progress>
    <p
      v-if="showValue"
      class="progress-value"
      :class="{ 'more-than-half': beyondHalf }">
      <slot>{{ newValue }}</slot>
    </p>
  </div>
</template>
