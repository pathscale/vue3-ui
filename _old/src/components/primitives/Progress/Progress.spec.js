import { shallowMount } from '@vue/test-utils'
import Progress from './Progress.vue'

let wrapper

describe('Progress', () => {
  beforeEach(() => {
    wrapper = shallowMount(Progress)
  })

  it('is called', () => {
    expect(wrapper.exists()).toBeTruthy()
  })

  it('render correctly', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })
})
