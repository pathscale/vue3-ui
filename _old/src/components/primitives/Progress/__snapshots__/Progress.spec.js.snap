// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Progress render correctly 1`] = `
<div class="progress-wrapper"><progress class="progress is-darkgrey" max="100"></progress>
  <!--v-if-->
</div>
`;
